git/gitattr/tree_test.go

```diff
@@ -0,0 +1,246 @@
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversSimpleTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("foo.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversSimpleChildrenTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("child/foo.dat")
+
+	assert.Len(t, attrs, 4)
+	assert.Equal(t, attrs[0], &Attr{K: "filter", V: "lfs"})
+	assert.Equal(t, attrs[1], &Attr{K: "diff", V: "lfs"})
+	assert.Equal(t, attrs[2], &Attr{K: "merge", V: "lfs"})
+	assert.Equal(t, attrs[3], &Attr{K: "text", V: "false"})
+}
+
+func TestNewDiscoversIndirectChildrenTrees(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "")
+	require.NoError(t, err)
+	defer os.Remove(tmp)
+
+	db, err := gitobj.FromFilesystem(tmp, "")
+	require.NoError(t, err)
+	defer db.Close()
+
+	blob, err := db.WriteBlob(gitobj.NewBlobFromBytes([]byte(`
+		*.dat filter=lfs diff=lfs merge=lfs -text
+	`)))
+	require.NoError(t, err)
+
+	indirect, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     ".gitattributes",
+			Oid:      blob,
+			Filemode: 0100644,
+		},
+	}})
+
+	child, err := db.WriteTree(&gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "indirect",
+			Oid:      indirect,
+			Filemode: 040000,
+		},
+	}})
+
+	tree, err := New(db, &gitobj.Tree{Entries: []*gitobj.TreeEntry{
+		{
+			Name:     "child",
+			Oid:      child,
+			Filemode: 040000,
+		},
+	}})
+	require.NoError(t, err)
+
+	attrs := tree.Applied("child/indirect/foo.dat")
```

Comment:
Same thing here as above - this matches on a top-level `*.dat` pattern, so it would be good to manually check the attributes and children of the tree itself

Reply:
Thanks, I caught this as well in https://github.com/git-lfs/git-lfs/commit/ff62456c555e0396b24585ec907fd56784469e1d.